File: cpp/src/TerseDecompresser.cpp
#include "TerseDecompresser.h"
#include "Constants.h"
#include "NonSpackDecompresser.h"
#include "SpackDecompresser.h"

#include <cstring>
#include <iostream>
#include <stdexcept>

std::unique_ptr< TerseDecompresser > TerseDecompresser::create(std::istream &inputStream, std::ostream &outputStream,
                                                               const DecompresserOptions &options)
{
  TerseHeader header = TerseHeader::parseHeader(inputStream);
  if (header.spackFlag)
  {
    return std::make_unique< SpackDecompresser >(inputStream, outputStream, header, options);
  }
  else
  {
    return std::make_unique< NonSpackDecompresser >(inputStream, outputStream, header, options);
  }
}

TerseDecompresser::~TerseDecompresser()
{
}

TerseDecompresser::TerseDecompresser(std::istream &in, std::ostream &out, const TerseHeader &header, const DecompresserOptions &options)
    : header_(header), options_(options), inputStream_(in), outputStream_(out)
{
  blockReader_ = std::make_unique< TerseBlockReader >(inputStream_);
  // Reserve enough space for the record buffer to avoid reallocations.
  // For variable-length records, we can have up to 32767 bytes.
  record_.reserve(header_.variableFlag ? 32767 : (header_.recordLength * 2));
}

void TerseDecompresser::putChar(int x)
{
  if (x == 0 && options_.textMode && header_.variableFlag == true)
  {
      endRecord();
  }
  else if (x == Tersedecompress::RECORDMARK && header_.variableFlag == true)
  {
    endRecord();
  }
  else
  {
    if (record_.capacity() == record_.size())
    {
      throw std::runtime_error("Record size exceeds expected maximum. File may be corrupted or not a valid Terse file.");
    }
    record_.push_back(options_.textMode ? static_cast< char >(Tersedecompress::EbcToAsc[x - 1])
                                        : static_cast< char >(x - 1));
    if (header_.variableFlag != true && static_cast< int >(record_.size()) == header_.recordLength)
    {
      endRecord();
    }
  }
}

void TerseDecompresser::endRecord()
{
  if (header_.variableFlag && !options_.textMode)
  {
    // Write a 4-byte Record Descriptor Word (RDW)
    // RDW = (record_.size() + 4) in high-order 2 bytes + low-order 2 bytes
    size_t lengthWithRDW = record_.size() + 4;
    // For MVS VB:
    //   bytes[0..1] = (16-bit) length of record
    //   bytes[2..3] = reserved (usually zero)
    unsigned char rdw[4];
    rdw[0] = static_cast< unsigned char >((lengthWithRDW >> 8) & 0xFF); // high-order length byte
    rdw[1] = static_cast< unsigned char >(lengthWithRDW & 0xFF);        // low-order length byte
    rdw[2] = 0x00;
    rdw[3] = 0x00;
    outputStream_.write(reinterpret_cast< const char * >(rdw), 4);
  }
  if (options_.textMode)
  {
    record_.push_back(0x0A);
  }
  if (record_.empty())
  {
    return;
  }
  {
    outputStream_.write(record_.data(), record_.size());
  }
  record_.clear();
}

void TerseDecompresser::close() 
{
  if (!record_.empty() || (options_.textMode && header_.variableFlag))
  {
    endRecord();
  }
  outputStream_.flush();
}

File: src/main/java/org/openmainframeproject/tersedecompress/TerseDecompresser.java
package org.openmainframeproject.tersedecompress;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;

abstract class TerseDecompresser implements AutoCloseable
{
	TerseBlockReader input;
	ByteBuffer record;
	DataOutputStream stream;
    final TerseHeader header;
    final boolean TextFlag;
    
    public abstract void decode() throws IOException;
    
    public static TerseDecompresser create(InputStream inputStream, OutputStream outputStream, boolean textMode) throws IOException
    {
        DataInputStream input = new DataInputStream(new BufferedInputStream(inputStream));
        TerseHeader header_rv = TerseHeader.CheckHeader(input);
        
        if (!header_rv.SpackFlag) {
        	return new NonSpackDecompresser(input, outputStream, header_rv, textMode);
        } else {
        	return new SpackDecompresser(input, outputStream, header_rv, textMode);
        }
    }
    
	public TerseDecompresser(InputStream instream, OutputStream outputStream, TerseHeader header, boolean textMode)
	{		
		this.header = header;
		this.TextFlag = textMode;
		this.input = new TerseBlockReader(instream);
		this.stream = new DataOutputStream(new BufferedOutputStream(outputStream));
		
        // Reserve enough space for the record buffer to avoid reallocations.
        // For variable-length records, we can have up to 32767 bytes.
		this.record = ByteBuffer.allocate(header.VariableFlag ? 32767 : (header.RecordLength * 2));
	}
	
    /* Write a new line to the output file*/
    public void endRecord() throws IOException 
    {
    	if (header.VariableFlag && !TextFlag)
    	{
    		// write a RDW
    		int recordlength = record.position() + 4;
    		int rdw = recordlength << 16;
    		stream.writeInt(rdw);
    	}
        if (TextFlag) {
            record.put((byte)0x0A);
        }
        stream.write(record.array(), 0, record.position());
        record.clear();
    }

    /*
     * Write some stuff to the output record
     */

    public void PutChar(int X) throws IOException {
        if (X == 0 && TextFlag && header.VariableFlag == true) {
            endRecord();
        } else if (X == Constants.RECORDMARK && header.VariableFlag == true) {
            endRecord();
        } else {
            if (record.capacity() == record.position()) {
                throw new IOException("Record size exceeds expected maximum. File may be corrupted or not a valid Terse file.");
            }
            record.put(TextFlag ? (byte) Constants.EbcToAsc[X - 1] : (byte) (X - 1));
            if (header.VariableFlag != true && (int) (record.position()) == header.RecordLength) {
                endRecord();
            }
        }
    }

	@Override
	public void close() throws Exception {
		if (record.position() > 0 || TextFlag && header.VariableFlag)
		{
			endRecord();
		}
		this.stream.close();
		this.input.close();
	}
}
